### PR TITLE
implement interp_like

### LINKED ISSUE #2218: interp_like
Just as a reminder of the remaining extension of #2104 .

We might add `interp_like` that behaves like `reindex_like` but using `interp()`.

### PR DESCRIPTION
**Pull request by fujiisoup**
- [x] Closes #2218
 - [x] Tests added
 - [x] Tests passed
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API.

This adds `interp_like`, that behaves like `reindex_like` but using interpolation.

### CODE REVIEW

**Comment by fujiisoup** [comment]
Slightly wondering the case when an object has non-numeric dimensions.
The current implementation raises TypeError, even when the target destination is just a subset of the current one.

**Comment by shoyer** [comment]
> Slightly wondering the case when an object has non-numeric dimensions.
The current implementation raises TypeError, even when the target destination is just a subset of the current one.

Do you have an example?

It might make sense to interpolate on numeric dimensions, and directly reindex on non-numeric dimensions.

**Comment by fujiisoup** [comment]
> Do you have an example?

We raise a TypeError in the following example, because the dataset has datetime-type dimension.

```python
In [3]: ds = xr.tutorial.load_dataset('air_temperature')

In [4]: ds
Out[4]: 
<xarray.Dataset>
Dimensions:  (lat: 25, lon: 53, time: 2920)
Coordinates:
  * lat      (lat) float32 75.0 72.5 70.0 67.5 65.0 62.5 60.0 57.5 55.0 52.5 ...
  * lon      (lon) float32 200.0 202.5 205.0 207.5 210.0 212.5 215.0 217.5 ...
  * time     (time) datetime64[ns] 2013-01-01 2013-01-01T06:00:00 ...
Data variables:
    air      (time, lat, lon) float32 241.2 242.5 243.5 244.0 244.09999 ...
Attributes:
    Conventions:  COARDS
    title:        4x daily NMC reanalysis (1948)
    description:  Data is from NMC initialized reanalysis\n(4x/day).  These a...
    platform:     Model
    references:   http://www.esrl.noaa.gov/psd/data/gridded/data.ncep.reanaly...

In [5]: ds.interp_like(ds.isel(lat=[0, 1, 2]))  # -> TypeError
```

> It might make sense to interpolate on numeric dimensions, and directly reindex on non-numeric dimensions.

I agree for the object type dimension, such as string, reindex is be appropriate and users might expect it.
But for datetime dimension, I think users expect the interpolation rather than reindexing.

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -1893,6 +1893,58 @@ def maybe_variable(obj, k):
...
+        numeric_coords = OrderedDict()
+        object_coords = OrderedDict()
+        for k, v in coords.items():
+            if v.dtype.kind in 'uifcMm':
+                numeric_coords[k] = v
+            else:
+                object_coords[k] = v
+        # We do not support interpolation along object coordinate.
+        # reindex instead.
+        ds = self.reindex(object_coords)
```
Comment (shoyer) [inline xarray/core/dataset.py]: Reindexing makes a copy of the dataset, so let's only do this if there are object coordinates.

File: xarray/core/missing.py [hunk 2]
```diff
@@ -395,6 +395,24 @@ def _localize(var, indexes_coords):
...
+    x = list(x)
+    new_x = list(new_x)
+    for i in range(len(x)):
+        if x[i].dtype.kind in 'Mm':
+            # Scipy cast np.float_ for coordinate, but 64bit float is not
+            # sufficient for datetime64 (uses 64bit integer).
+            # We assumes x - min(x) can be well presented by 64bit float.
+            xmin = np.min(x[i])
+            x[i] = (x[i] - xmin).astype(np.float_)
+            new_x[i] = (new_x[i] - xmin).astype(float)
```
Comment (shoyer) [inline xarray/core/missing.py]: Can we use the same `astype(np.float64)` on both these lines?

File: xarray/tests/test_interp.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 from __future__ import absolute_import, division, print_function
 
 import numpy as np
+import pandas as pd
```
[the remark this answers is not in the document]
Comment (shoyer) [reply xarray/tests/test_interp.py]: looks like we can remove this?

File: xarray/core/dataarray.py [hunk 4]
```diff
@@ -951,6 +951,52 @@ def interp(self, coords=None, method='linear', assume_sorted=False,
             **coords_kwargs)
         return self._from_temp_dataset(ds)
 
+    def interp_like(self, other, method='linear', assume_sorted=False,
+                    kwargs={}):
+        """Interpolate this object onto the coordinates of another object,
+        filling out of range values with NaN.
+
```
Comment (shoyer) [inline xarray/core/dataarray.py]: Can we add a note that if there are coordinates with object dtype, these are reindexed instead, not interpolated?

**Comment by fujiisoup** [comment]
Thanks for the review :) merged.

**Pull request merged**